Getting StringIndexOutOfBoundsException on textDocument/signatureHelp request when triggered from the end of document

### Issue
When we trigger a `textDocument/signatureHelp` request from the end of the document, it throws `java.lang.StringIndexOutOfBoundsException` as the offset exceeds the buffer length.

### How to Reproduce
Trigger a `textDocument/signatureHelp` request from the end of the document (ensure to write both open & close parenthesis manually).

### Stack trace
```
java.util.concurrent.CompletionException: java.lang.StringIndexOutOfBoundsException: String index out of range: 6
  at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
  at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
  at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
  at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
  at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
  at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
  at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
  at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
  at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 6
  at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:48)
  at java.base/java.lang.String.charAt(String.java:1513)
  at org.eclipse.jdt.ls.core.internal.handlers.SignatureHelpContext.findTargetNode(SignatureHelpContext.java:132)
  at org.eclipse.jdt.ls.core.internal.handlers.SignatureHelpContext.resolve(SignatureHelpContext.java:103)
  at org.eclipse.jdt.ls.core.internal.handlers.SignatureHelpUtils.getSignatureHelpFromASTNode(SignatureHelpUtils.java:52)
  at org.eclipse.jdt.ls.core.internal.handlers.SignatureHelpHandler.signatureHelp(SignatureHelpHandler.java:79)
  at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$8(JDTLanguageServer.java:649)
  at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:87)
  at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
... 6 more
```

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SignatureHelpContext.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SignatureHelpContext.java
@@ -123,7 +123,7 @@ public class SignatureHelpContext {
 		}
 
 		String source = unit.getSource();
-		if (source == null) {
+		if (source == null || triggerOffset >= source.length()) {
 			return;
 		}
 

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/SignatureHelpHandlerTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/SignatureHelpHandlerTest.java
@@ -181,6 +181,20 @@ public class SignatureHelpHandlerTest extends AbstractCompilationUnitBasedTest {
 		assertEquals(0, help.getSignatures().size());
 	}
 
+	@Test
+	public void testSignatureHelp_endOfDoc() throws JavaModelException {
+		IPackageFragment pack1 = sourceFolder.createPackageFragment("test1", false, null);
+		StringBuilder buf = new StringBuilder();
+		buf.append("package test1;\n");
+		buf.append("public class E {\n");
+		buf.append("   public int bar(String s) {  }\n");
+		buf.append("}\n");
+		ICompilationUnit cu = pack1.createCompilationUnit("E.java", buf.toString(), false, null);
+		SignatureHelp help = getSignatureHelp(cu, 3, 2);
+		assertNotNull(help);
+		assertEquals(0, help.getSignatures().size());
+	}
+
 	// See https://github.com/eclipse/eclipse.jdt.ls/pull/1015#issuecomment-487997215
 	@Test
 	public void testSignatureHelp_parameters() throws JavaModelException {

